File: src/main/java/frc/robot/constants/ClimberConstants.java
package frc.robot.constants;

import static edu.wpi.first.units.Units.Amps;
import static edu.wpi.first.units.Units.Inches;
import static edu.wpi.first.units.Units.Meters;

import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.Distance;
import frc.utils.controlWrappers.PIDGains;

public class ClimberConstants {
    public static final int CLIMBER_MOTOR_ID = 5;

    public static final boolean CLIMBER_INVERTED = false;
    public static final Current CLIMBER_CURRENT_LIM = Amps.of(30);

    public static final Distance CLIMBER_EXTEND_POSITION = Meters.of(1);
    public static final Distance CLIMBER_RETRACT_POSITION = Meters.of(0);

    public static final Distance CLIMBER_SETPOINT_TOLERANCE = Inches.of(1);

    public static final PIDGains.GravityFF CLIMBER_ID_GAINS = new PIDGains.GravityFF(0.1, 0, 0.1, 0.0001);

    public static final PIDGains.GravityFF CLIMBER_FF_GAINS = new PIDGains.GravityFF(0.1, 0, 0.1, 0).makeTunable("Tuning/Climber/FF");
    public static final PIDGains.ProfiledPID CLIMBER_PID_GAINS = new PIDGains.ProfiledPID(1, 1, 1, 100, 100).makeTunable("Tuning/Climber/PID");

    private static final Distance CLIMBER_SPOOL_RADIUS = Inches.of(1); //TODO: get actual radius later
    private static final double CLIMBER_GEAR_RATIO = 80.0;
    public static final double CLIMBER_POSITION_CONVERSION_FACTOR = (CLIMBER_SPOOL_RADIUS.in(Meters)*2*Math.PI)/CLIMBER_GEAR_RATIO;
    public static final double CLIMBER_VELOCITY_CONVERSION_FACTOR = CLIMBER_POSITION_CONVERSION_FACTOR;
}

File: src/main/kotlin/frc/robot/constants/ClimberConstants.kt
package frc.robot.constants

import edu.wpi.first.units.Units.Amps
import edu.wpi.first.units.Units.Inches
import edu.wpi.first.units.Units.Meters
import edu.wpi.first.units.Units.MetersPerSecond
import edu.wpi.first.units.Units.Seconds
import edu.wpi.first.units.Units.Volts
import edu.wpi.first.units.measure.Current
import edu.wpi.first.units.measure.Distance
import edu.wpi.first.units.measure.LinearVelocity
import edu.wpi.first.units.measure.Time
import edu.wpi.first.units.measure.Voltage
import frc.utils.controlWrappers.PIDGains

object ClimberConstants {
    const val CLIMBER_MOTOR_ID: Int = 5

    const val CLIMBER_INVERTED: Boolean = false
    @JvmField val CLIMBER_CURRENT_LIM: Current = Amps.of(30.0)

    @JvmField val CLIMBER_MAX_POSITION: Distance = Inches.of(16.0)
    @JvmField val CLIMBER_MIN_POSITION: Distance = Meters.of(0.0)

    @JvmField val CLIMBER_SETPOINT_TOLERANCE: Distance = Inches.of(1.0)

    @JvmField val CLIMBER_ID_GAINS: PIDGains.GravityFF = PIDGains.GravityFF(0.1, 0.0, 0.1, 0.0001)

    @JvmField val CLIMBER_FF_GAINS: PIDGains.GravityFF = PIDGains.GravityFF(0.1, 0.0, 0.1, 0.0).makeTunable("Tuning/Climber/FF")
    @JvmField val CLIMBER_PID_GAINS: PIDGains.ProfiledPID = PIDGains.ProfiledPID(0.1, 0.0, 0.0, 10.0, 100.0).makeTunable("Tuning/Climber/PID")

    private val CLIMBER_SPOOL_RADIUS: Distance = Inches.of(1.0)
    private const val CLIMBER_GEAR_RATIO: Double = 80.0
    @JvmField val CLIMBER_POSITION_CONVERSION_FACTOR: Double = (CLIMBER_SPOOL_RADIUS.`in`(Meters) * 2 * Math.PI) / CLIMBER_GEAR_RATIO
    @JvmField val CLIMBER_VELOCITY_CONVERSION_FACTOR: Double = CLIMBER_POSITION_CONVERSION_FACTOR

    const val CLIMBER_HOME_ON_START: Boolean = false
    @JvmField val CLIMBER_HOME_VOLTAGE: Voltage = Volts.of(-2.0)
    @JvmField val CLIMBER_HOME_STOP_TIME: Time = Seconds.of(0.5)
    @JvmField val CLIMBER_HOME_STOP_THRESH: LinearVelocity = MetersPerSecond.of(0.01)
}
